x-pack/metricbeat/module/aws/ec2/ec2.go

```diff
@@ -0,0 +1,330 @@
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/aws"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet(aws.ModuleName, "ec2", New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MockEC2Client struct is used for unit tests.
+type MockEC2Client struct {
+	ec2iface.EC2API
+}
+
+// MockCloudWatchClient struct is used for unit tests.
+type MockCloudWatchClient struct {
+	cloudwatchiface.CloudWatchAPI
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	*aws.MetricSet
+	config *aws.Config
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The aws ec2 metricset is experimental.")
+
+	config := aws.Config{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	os.Setenv("AWS_ACCESS_KEY_ID", config.AccessKeyID)
+	os.Setenv("AWS_SECRET_ACCESS_KEY", config.SecretAccessKey)
+	os.Setenv("AWS_SESSION_TOKEN", config.SessionToken)
+
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating aws metricset")
+	}
+
+	return &MetricSet{
+		MetricSet: metricSet,
+		config:    &config,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	//mock Fetch
+	if m.config.Mock == "true" {
+		m.MockFetch(report)
+		return
+	}
+
+	//actual fetch function
+	cfg, err := external.LoadDefaultAWSConfig()
+	if err != nil {
+		fmt.Println("Failed to load config: ", err.Error())
+	}
+
+	cfg.Region = "us-west-1"
+	svcEC2 := ec2.New(cfg)
+	//Get a list of regions
+	regionsList, err := getRegions(svcEC2)
+	if err != nil {
+		report.Error(errors.Wrap(err, "getRegions failed"))
+	}
+
+	for _, regionName := range regionsList {
+		cfg.Region = regionName
+		svcEC2 := ec2.New(cfg)
+		instanceIDs, instancesOutputs, err := getInstancesPerRegion(svcEC2)
+		if err != nil {
+			report.Error(errors.Wrap(err, "getInstancesPerRegion failed"))
```

Comment:
`report.Error` won't stop this method, do we really want to continue with the other regions? at the very least we should call `continue` to skip this one

Reply:
Yeah good catch on the `continue`. I do think we want to continue with other regions though. For example I had seen regionList includes 'us-gov' before and it will return an error here since we dont have permission to visit us-gov. But the same api call works fine with other regions.